model/userdata.go

```diff
@@ -147,3 +148,105 @@ func validateCoreosCloudInit(content []byte) error {
 func validateNone([]byte) error {
 	return nil
 }
+
+func (self UserDataPart) Base64UserDataRef() (string, error) {
+	// Change this to Base64IgnitionConfigRef to test with ignition
+	return self.Base64BashScriptRef()
+}
+
+func (self UserDataPart) Base64IgnitionConfigRef() (string, error) {
+	base64Part, err := self.Base64(false)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate ignition config: %v", err)
+	}
+
+	// TODO Append user-provided ignition config via `ignition.config.append.source`
+	// once it supports sourcing from a s3 object
+	// https://coreos.com/ignition/docs/0.14.0/configuration-v2_0.html
+
+	head, err := json.Marshal(`{
+  "ignition": {
+    "version": "2.0.0"
+  },
+  "storage": {
+    "files": [
+      {
+        "filesystem": "root",
+        "path": "/kube-aws-stack-name",
+        "mode": 420,
+        "contents": {
+          "source": "data:;base64,`)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate ignition config: %v", err)
+	}
+
+	tail, err := json.Marshal(
+		fmt.Sprintf(`"
+        }
+      },
+      {
+        "filesystem": "root",
+        "path": "/opt/bin/bootstrap-worker-node",
+        "mode": 493,
+        "contents": {
+          "source": "data:;base64,%s"
+        }
+      }
+    ]
+  },
+  "systemd": {
+    "units": [
+      {
+        "name": "bootstrap-worker-node.service",
+        "enable": true,
+        "contents": "[Unit]\nRequires=network-online.target\nAfter=network-online.target\n\n[Service]\nType=oneshot\nRemainAfterExit=yes\nExecStart=/opt/bin/bootstrap-worker-node\n\n[Install]\nWantedBy=multi-user.target"
+      }
+    ]
+  }
+}`,
+			base64Part,
+		),
+	)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate ignition config: %v", err)
+	}
+
+	splits := []string{
+		string(head),
+		`{"Fn::Base64":{"Ref":"AWS::StackName"}}`,
+		string(tail),
+	}
+
+	// TODO should emit a validation error when sum(sizeOf(split in splits)) > 16384
+
+	return fmt.Sprintf(`{"Fn::Base64": {"Fn::Join": ["", [%s]]} }`, strings.Join(splits, ",")), nil
+}
+
+func (self UserDataPart) Base64BashScriptRef() (string, error) {
+	cloudConfig, err := self.Template()
+	if err != nil {
+		return "", fmt.Errorf("failed to generate userdata: %v", err)
+	}
+
+	cloudConfigLines := strings.Split(cloudConfig, "\n")
+
+	jsonStrings := []string{}
+	for _, o := range cloudConfigLines {
+		escaped, err := json.Marshal(o + "\n")
+		if err != nil {
+			return "", fmt.Errorf("failed to generate userdata: failed to marshal \"%s\" into json: %v", o, err)
+		}
+		jsonStrings = append(jsonStrings, string(escaped))
+	}
+
+	splits := []string{}
+	splits = append(splits, jsonStrings[:1]...)
+	splits = append(splits, `"echo \""`)
+	splits = append(splits, `{"Ref":"AWS::StackName"}`)
+	splits = append(splits, `"\" >> /kube-aws-stack-name\n"`)
```

Comment:
This func is used to inject `{"Ref":"AWS::StackName"}` into nodes by modifying the "instance" parts of userdata.